src/execution/BucketExecutor.tsx

```diff
@@ -0,0 +1,338 @@
+                if (measureDimension === null) {
+                    measureDimension = dimensionIndex;
+                    itemIdentifiers.push(MEASUREGROUP);
+                } else if (measureDimension !== dimensionIndex) {
+                    throw new RuntimeError(
+                        `All measures have to be on the same dimension. There are some measures already on dimension ${measureDimension} but measure ${measureLocalIdentifier} is on dimension ${dimensionIndex}.`,
+                    );
+                }
+                measuresBucket.items.push(dimensionItem);
+            }
+            resultSpec.dimensions.push({
+                itemIdentifiers,
+            });
+        });
+    });
+
+    const buckets: VisualizationObject.IBucket[] = [];
+    if (attributesBucket.items.length > 0) {
+        buckets.push(attributesBucket);
+    }
+    if (measuresBucket.items.length > 0) {
+        buckets.push(measuresBucket);
+    }
+    if (buckets.length === 0) {
+        throw new RuntimeError("No measures nor attributes defined");
+    }
+
+    const afm = convertBucketsToAFM(buckets, filters);
+
+    return {
+        execution: {
+            afm,
+            resultSpec,
+        },
+    };
+};
+
+export interface IPaging {
+    offset: number[];
+    limit: number[];
+}
+
+export interface IBucketExecutorChildrenProps {
+    isLoading: boolean;
+    error: any;
+    result: Execution.IExecutionResult | null;
+    response: Execution.IExecutionResponse | null;
+    getPage: (paging: IPaging) => void;
+}
+
+export interface IBucketExecutorProps {
+    telemetryComponentName?: string;
+    autoLoadFirstPage?: boolean;
+    initialPaging?: IPaging;
+    sdk?: SDK;
+
+    children: (childrenProps: IBucketExecutorChildrenProps) => React.ReactNode;
+    dimensions: dimensionItem[][];
+    projectId: string;
+    filters?: VisualizationInput.IFilter[];
+    sortBy?: VisualizationInput.ISort[];
+}
+
+export interface IBucketExecutorState {
+    result: Execution.IExecutionResult | null;
+    response: Execution.IExecutionResponse | null;
+    isLoading: boolean;
+    error: any;
+    responsePromise: Promise<Execution.IExecutionResponse> | null;
+    resultPromise: Promise<Execution.IExecutionResult> | null;
+
+    children: IBucketExecutorProps["children"];
+    dimensions: IBucketExecutorProps["dimensions"];
+    projectId: IBucketExecutorProps["projectId"];
+    filters: IBucketExecutorProps["filters"];
+    sortBy: IBucketExecutorProps["sortBy"];
+}
+
+export interface IBucketExecutorChildrenProps {
+    response: Execution.IExecutionResponse | null;
+    result: Execution.IExecutionResult | null;
+    isLoading: boolean;
+    error: any;
+    getPage: (paging: IPaging) => void;
+}
+
+const propsRequiringDataReset: Array<keyof IBucketExecutorProps> = [
+    "projectId",
+    "dimensions",
+    "filters",
+    "sortBy",
+];
+
+const didPropsChange = <V extends {}>(propList: string[], prevProps: V, props: V): boolean => {
```

Comment:
`propList` can hopefully typed as `Array<keyof IBucketExecutorProps>` if you make `V extends IBucketExecutorProps` which it probably does if I understand this correctly. Moreover, this function is called from only one place, so its types are known -> there is no reason for it to be generic.